llvm/lib/Target/NVPTX/NVPTXTargetTransformInfo.h

```diff
@@ -83,6 +83,8 @@ class NVPTXTTIImpl : public BasicTTIImplBase<NVPTXTTIImpl> {
   }
   unsigned getMinVectorRegisterBitWidth() const { return 32; }
 
+  bool shouldExpandReduction(const IntrinsicInst *II) const { return false; }
```

Comment:
Why have we completely disabled the generic expansion? It seems like for many of these operations we're not doing anything special and the existing SelectionDAG support would work just as well.

Reply:
This line only disables expansion in the IR-level `ExpandReductions` pass, letting SelectionDAG handle it instead. There are different behaviors between the IR-level pass, DAGTypeLegalizer, and the implementation in NVPTX being considered:

- ExpandReductions uses shufflevectors to split the vector operand into two smaller vectors, partially reduce them, and so on.
- SelectionDAG achieves the same thing without using shufflevectors.
- This implementation groups nearby elements bottom-up. I think this is better for data dependences.

Reply:
To rephrase:

There are three types of reductions possible: sequential (`a + b + c + d`), shuffle (`(a + c) + (b + d)`), and tree (`(a + b) + (c + d)`). `ExpandReductions` pass and SelectionDAG only support the first two. If we handle reductions in the NVPTX backend, we can support all three.